### PR TITLE
Fix: Tooltip and message date position

### PR DESCRIPTION
**Pull request by rdjuric**
### Details

1. Style changes to make the Tooltip responsive instead of fix width.
2. Apply the Tooltip containerStyle in index.native.js for consistency across platforms
3. Fixes the message timestamp to the same line as the displayName

### Fixed Issues

Fixes [an issue with the below PR ](https://github.com/Expensify/Expensify.cash/pull/3830#issuecomment-872556352) where the Tooltip and the Timestamp of a message would me misplaced.
$ PR #3830 

### Tests

1. Open any chat
2. Check that the timestamp is line with the displayName
3. Hoover over the displayName
4. The Tooltip should show close to your cursor

### QA Steps

Same as above.
### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/53711423/125451787-291abf84-df6d-4de2-a7e2-4b577aef5220.mov

#### Mobile Web

https://user-images.githubusercontent.com/53711423/125451822-1b42ce10-4169-49e3-a811-f310c0b4bb3b.mp4

#### Desktop

https://user-images.githubusercontent.com/53711423/125451839-824bee7e-cab9-4be4-a334-840f298ffbdc.mov

#### iOS

https://user-images.githubusercontent.com/53711423/125451846-375d8db5-0714-4352-9fa7-cab7fb327130.mp4

#### Android

https://user-images.githubusercontent.com/53711423/125451860-1cf24716-b4d6-40f4-8f86-d09898cb0b87.mp4

### CODE REVIEW

**Comment by Luke9389** [comment]
@rdjuric in the future, with regressions, feel free to continue to tag and request reviews from engineers that are connected to the original issue

**Comment by rdjuric** [comment]
> @rdjuric in the future, with regressions, feel free to continue to tag and request reviews from engineers that are connected to the original issue

Makes sense! Will do that in the future @Luke9389

File: src/pages/home/report/ReportActionItemFragment.js [hunk 1]
```diff
@@ -83,7 +83,7 @@ class ReportActionItemFragment extends React.PureComponent {
                     );
             case 'TEXT':
                 return (
-                    <Tooltip text={tooltipText} containerStyle={styles.w100}>
```
Comment (Luke9389) [inline src/pages/home/report/ReportActionItemFragment.js:86]: Would it work for us to pass a prop here rather than putting a view wrapper an all of our tooltip components?

Comment (rdjuric) [reply src/pages/home/report/ReportActionItemFragment.js:86]: Our `Tooltip` for the web already had a view wrapper, but our native version was just returning the children without any container (and consequently without the containerStyle). This PR is just adding a view wrapper to the native `Tooltip` too so we keep the styles consistent between native and web.

We use the `View` wrapper in the web version to [measure and position ](https://github.com/Expensify/Expensify.cash/blob/2f2f4c484ca42c7b1ec27c7df380cbe2e97e4a57/src/components/Tooltip/index.js#L84)the `Tooltip`, so removing it doesn't sound good.

**Comment by Luke9389** [comment]
linking this to the original issue: https://github.com/Expensify/Expensify.cash/issues/3172

File: src/pages/home/report/ReportActionItemFragment.js (hunk 1 above)
Comment (Luke9389) [reply src/pages/home/report/ReportActionItemFragment.js:86]: Sorry, let me rephrase that question. Can we accomplish this fix without making the switch to `flexShrink1`? The original issue just added `styles.w100`. What I'm asking is if we can keep that approach, but just make it something we can turn on and off with a prop. Does that make sense?

Comment (rdjuric) [reply src/pages/home/report/ReportActionItemFragment.js:86]: Ah, I see! I don't think we can keep the `styles.w100` without causing the Tooltip to be misplaced, and flexShrink1 makes more sense here since it's specific to the issue at hand (i.e shrinking a component when it gets bigger than its flex container).

Would you prefer if the `styles.flexShrink1` was turned on by the `isSingleLine` prop? Something like
```
<Tooltip text={tooltipText} containerStyle={this.props.isSingleLine ? styles.flexShrink1 : undefined}>
```

File: src/components/Tooltip/index.native.js [hunk 2]
```diff
@@ -1,18 +1,37 @@
...
+import {View} from 'react-native';
 import PropTypes from 'prop-types';
 
 // We can't use the common component for the Tooltip as Web implementation uses DOM specific method to
 // render the View which is not present on the Mobile.
 const propTypes = {
-    children: PropTypes.element,
+    /** Styles to be assigned to the Tooltip wrapper views */
+    // eslint-disable-next-line react/forbid-prop-types
+    containerStyle: PropTypes.object,
```
Comment (Luke9389) [inline src/components/Tooltip/index.native.js:10]: I think we prefer the styles to be `PropTypes.arrayOf(PropTypes.object)`

File: src/pages/home/report/ReportActionItemFragment.js (hunk 1 above)
Comment (Luke9389) [reply src/pages/home/report/ReportActionItemFragment.js:86]: Ok, gotcha. I was trying to think of ways to avoid the wrapper, but this is definitely a clear way of doing it.

File: src/components/Tooltip/index.native.js [hunk 3]
```diff
@@ -1,18 +1,37 @@
...
+
+    /** Children to wrap with Tooltip. */
+    children: PropTypes.node.isRequired,
+};
+
+const defaultProps = {
+    containerStyle: {},
```
Comment (Luke9389) [inline src/components/Tooltip/index.native.js:17]: This should be `[]`

File: src/components/Tooltip/index.native.js (hunk 2 above)
Comment (Luke9389) [reply src/components/Tooltip/index.native.js:10]: and we can rename this to be `containerStyles`

Comment (rdjuric) [reply src/components/Tooltip/index.native.js:10]: Ok! I'll change it on the non-native version too and adapt our current `Tooltip`s to the change

I guess it's better to change the `Hooverable` too since it's also receiving an object as the style?

**Comment by rdjuric** [comment]
Update made @Luke9389! 

Changed our `containerStyles` to receive an **_array_** in our `Tooltip` and `Hoverable` to keep styling consistent across our code. Nice catch, thanks.

File: src/components/DisplayNames/index.js [hunk 4]
```diff
@@ -95,7 +95,7 @@ class DisplayNames extends PureComponent {
                             <Tooltip
                                 key={index}
                                 text={tooltip}
-                                containerStyle={styles.dInline}
+                                containerStyles={[styles.dInline]}
```
Comment (parasharrajat) [inline src/components/DisplayNames/index.js:98]:
```suggestion
                                containerStyle={[styles.dInline]}
```
`containerStyle` is more appropriate. There are no multiple styles. Its fine if it takes array.

File: src/components/Tooltip/index.native.js [hunk 5]
```diff
@@ -1,18 +1,37 @@
...
- * There is no native support for the Hover on the Mobile platform so we just return the enclosing childrens
+ * There is no native support for the Hover on the Mobile platform, but as we use the Tooltip as a
+ * container we must past pass that containerStyle to a simple View in order to avoid different
+ * styles across platforms.
  * @param {propTypes} props
  * @returns {ReactNodeLike}
  */
-const Tooltip = props => props.children;
+const Tooltip = props => (
+    <View style={[props.containerStyles]}>
```
Comment (parasharrajat) [inline src/components/Tooltip/index.native.js:28]:
```suggestion
    <View style={props.containerStyles}>
```

File: src/components/Tooltip/index.native.js [hunk 6]
```diff
@@ -1,18 +1,37 @@
...
+        {props.children}
+    </View>
+);
+
 
 Tooltip.propTypes = propTypes;
+Tooltip.defaultProps = defaultProps;
 Tooltip.displayName = 'Tooltip';
 export default Tooltip;
```
Comment (parasharrajat) [inline src/components/Tooltip/index.native.js:37]: Please Add an empty line over it.

File: src/components/Tooltip/index.native.js [hunk 7]
```diff
@@ -1,18 +1,37 @@
...
  * @param {propTypes} props
  * @returns {ReactNodeLike}
  */
-const Tooltip = props => props.children;
+const Tooltip = props => (
+    <View style={[props.containerStyles]}>
+        {props.children}
+    </View>
+);
+
```
Comment (parasharrajat) [inline src/components/Tooltip/index.native.js:32]: Extra line

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/DisplayNames/index.js (hunk 4 above)
Comment (rdjuric) [reply src/components/DisplayNames/index.js:98]: I think `containerStyles` is better because
- It makes it clearer that we're taking multiple styles (array)
- Is consistent with the rest of the code (we have more components with a `containerStyles` prop)
- Even if we're not adding multiple styles in this specific use of the Tooltip, we might want to it in the future

File: src/components/Tooltip/index.native.js (hunk 5 above)
Comment (rdjuric) [reply src/components/Tooltip/index.native.js:28]: Good catch!

**Comment by rdjuric** [comment]
Made a few changes after your comments @parasharrajat. Thanks!

File: src/components/DisplayNames/index.js (hunk 4 above)
Comment (parasharrajat) [reply src/components/DisplayNames/index.js:98]: Thanks for clarification. I just checked the style guide. It change so dramatically. LOL. You are correct here.

File: src/components/Tooltip/index.native.js [hunk 8]
```diff
@@ -1,18 +1,36 @@
+import React from 'react';
+import {View} from 'react-native';
 import PropTypes from 'prop-types';
 
 // We can't use the common component for the Tooltip as Web implementation uses DOM specific method to
 // render the View which is not present on the Mobile.
 const propTypes = {
-    children: PropTypes.element,
+    /** Styles to be assigned to the Tooltip wrapper views */
+    // eslint-disable-next-line react/forbid-prop-types
```
Comment (Luke9389) [inline src/components/Tooltip/index.native.js:9]: I don't think we need this eslint-disable statement anymore, right?

**Comment by rdjuric** [comment]
Updated @Luke9389

File: src/components/Tooltip/index.native.js (hunk 2 above)
Comment (Luke9389) [reply src/components/Tooltip/index.native.js]: Yup, thanks

File: src/components/Tooltip/index.native.js [hunk 9]
```diff
@@ -1,18 +1,35 @@
...
+
+const defaultProps = {
+    containerStyles: [],
 };
 
 /**
- * There is no native support for the Hover on the Mobile platform so we just return the enclosing childrens
+ * There is no native support for the Hover on the Mobile platform, but as we use the Tooltip as a
+ * container we must past pass that containerStyle to a simple View in order to avoid different
+ * styles across platforms.
```
Comment (Luke9389) [inline src/components/Tooltip/index.native.js:22]: I don't think we need this comment here at all.

**Review by Luke9389 — CHANGES_REQUESTED** [review]
Just a few more small things

File: src/components/Tooltip/index.native.js [hunk 10]
```diff
@@ -1,18 +1,35 @@
...
 
 // We can't use the common component for the Tooltip as Web implementation uses DOM specific method to
 // render the View which is not present on the Mobile.
 const propTypes = {
-    children: PropTypes.element,
+    /** Styles to be assigned to the Tooltip wrapper views */
+    containerStyles: PropTypes.arrayOf(PropTypes.object),
+
+    /** Children to wrap with Tooltip. */
+    children: PropTypes.node.isRequired,
```
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/components/Tooltip/index.native.js:12]: Oh, I think that was a miss. At that time main focus was the Web.

**Comment by rdjuric** [comment]
Updated @Luke9389

**Comment by Luke9389** [comment]
OK, this is lookin' good. Can you be sure to retest on all 5 platforms with the most recent version of the code before we wrap this up?

**Comment by rdjuric** [comment]
> OK, this is lookin' good. Can you be sure to retest on all 5 platforms with the most recent version of the code before we wrap this up?

Will start retesting and tag you here once it's done!

**Comment by Luke9389** [comment]
Thanks! (I know it's a pain but I've seen looooots of regressions caused by skipping this final step)

**Comment by rdjuric** [comment]
@Luke9389

I tested this branch and the PR worked correctly. I merged it into our main branch locally just to make **_sure_** things were okay and looks like this commit https://github.com/Expensify/Expensify.cash/commit/7729fc0accab4011d3c6aacb1cdac2091c394e55 had a regression - it nullified the styling changes we made on the previous PR https://github.com/Expensify/Expensify.cash/pull/3830.

That's why we have this Tooltip/message date bug on prod (where the avatar and the display name aren't wrapped in the `Pressable`) but not on staging (where they're wrapped in a `Pressable`)

To make our change work with the `Pressable` we need to move our `flexShrink1` from the `ToolTip` to the `Pressable` component. The updated videos on the PR are from the latest commit, where I made this change.

I kept our `containerStyle` -> `containerStyles` changes even thought we're no longer passing any styling to the `ToolTip` because I think they make the code more consistent, but let me know if you think differently.

**Comment by Luke9389** [comment]
OK, that sounds fine to me 👍

File: src/components/Tooltip/index.native.js (hunk 7 above)
Comment (Luke9389) [reply src/components/Tooltip/index.native.js:28]: did this get resolved?

File: src/components/Tooltip/index.native.js (hunk 6 above)
Comment (Luke9389) [reply src/components/Tooltip/index.native.js:32]: did this get resolved?

**Comment by rdjuric** [comment]
Can we merge this? @Luke9389

**Review by Luke9389 — APPROVED** [review]

**Comment by Luke9389** [comment]
was just testing on all platforms myself as well :)

**Comment by rdjuric** [comment]
Thanks, @Luke9389! Sorry for missing this regression on the first PR.
